contracts/payment/SponsorPayment.sol

```diff
@@ -0,0 +1,519 @@
+     * @param token The address of the token (use address(0) for Ether)
+     */
+    function requestWithdrawalERC20(uint256 amount, address token) external validToken(token) {
+        uint256 lockTime = _requestWithdrawal(token, amount);
+        emit ERC20WithdrawalRequested(_msgSender(), token, amount, lockTime);
+    }
+
+    function _cancelWithdrawal(address token) private returns (uint256) {
+        SponsorPaymentStorage storage $ = _getSponsorPaymentStorage();
+        WithdrawalRequest storage request = $.withdrawalRequests[_msgSender()][token];
+
+        if (!request.exists) revert InvalidWithdraw("No withdrawal request exists");
+
+        uint256 amount = request.amount;
+        delete $.withdrawalRequests[_msgSender()][token];
+
+        // Return the funds
+        $.balances[_msgSender()][token] += amount;
+
+        return amount;
+    }
+
+    /**
+     * @notice Cancel a pending withdrawal request
+     */
+    function cancelWithdrawal() external {
+        uint256 amount = _cancelWithdrawal(address(0));
+        emit WithdrawalCancelled(_msgSender(), amount);
+    }
+
+    /**
+     * @notice Cancel a pending withdrawal request
+     * @param token The token address
+     */
+    function cancelWithdrawalERC20(address token) external validToken(token) {
+        uint256 amount = _cancelWithdrawal(token);
+        emit ERC20WithdrawalCancelled(_msgSender(), token, amount);
+    }
+
+    function _getWithdrawalAmount(address token) private view returns (uint256) {
+        SponsorPaymentStorage storage $ = _getSponsorPaymentStorage();
+        WithdrawalRequest memory request = $.withdrawalRequests[_msgSender()][token];
+
+        if (!request.exists) revert InvalidWithdraw("No withdrawal request exists");
+        if (block.timestamp < request.lockTime)
+            revert InvalidWithdraw("Withdrawal is still locked");
+
+        return request.amount;
+    }
+
+    /**
+     * @notice Execute withdrawal after delay period
+     */
+    function withdraw() external nonReentrant {
+        SponsorPaymentStorage storage $ = _getSponsorPaymentStorage();
+        uint256 amount = _getWithdrawalAmount(address(0));
+        delete $.withdrawalRequests[_msgSender()][address(0)];
+        (bool success, ) = payable(_msgSender()).call{value: amount}("");
+        if (!success) revert InvalidWithdraw("Transfer failed");
+        emit Withdrawal(_msgSender(), amount);
+    }
+
+    /**
+     * @notice Execute withdrawal after delay period
+     * @param token The address of the token
+     */
+    function withdrawERC20(address token) external nonReentrant validToken(token) {
+        SponsorPaymentStorage storage $ = _getSponsorPaymentStorage();
+        uint256 amount = _getWithdrawalAmount(token);
+        IERC20(token).safeTransfer(_msgSender(), amount);
+        delete $.withdrawalRequests[_msgSender()][token];
+        emit ERC20Withdrawal(_msgSender(), token, amount);
+    }
+
+    function _claimPayment(
```

Comment:
this method name confuses me.

Reply:
Please suggest the name that don't make you confuse

Reply:
discussion in the separate thread. I'm against of the usage the word 'claim' to avoid missunderstanding.